src/test/mod.rs

```diff
@@ -114,18 +115,28 @@ impl TestEnvironment {
 
 pub(crate) struct TestDatabase {
     conn: Arc<Mutex<Connection>>,
+    schema: String,
 }
 
 impl TestDatabase {
     fn new() -> Result<Self, Error> {
-        // The temporary migration uses CREATE TEMPORARY TABLE instead of CREATE TABLE, creating
-        // fresh temporary copies of the database on top of the real one. The temporary tables are
-        // only visible to this connection, and will be deleted when it exits.
+        // A random schema name is generated and used for the current connection. This allows each
+        // test to create a fresh instance of the database to run within.
+        let schema = format!("docs_rs_test_schema_{}", rand::random::<u64>());
+
         let conn = crate::db::connect_db()?;
-        crate::db::migrate_temporary(None, &conn)?;
+        conn.batch_execute(&format!(
+            "
+                CREATE SCHEMA {0};
+                SET search_path TO {0};
```

Comment:
I was curious whether there was a way for postgres to do this automatically - turns out there is!

```suggestion
                SET search_path TO pg_temp;
```

This says it's different for every connection, so I'm not sure if the tests will have difference schemas, but I think it's worth a shot :)

https://dba.stackexchange.com/a/76661

Reply:
You cannot look up functions inside `pg_temp`. I spent quite a while hacking around with it to see if I could get it working, but as far as I can tell it's impossible to use functions inside it without using an explicit schema-name when calling them.

Reply:
If I add `pg_temp` as a prefix it seems to work:

```
cratesfyi=> set search_path to pg_temp;
SET
cratesfyi=>                 CREATE FUNCTION normalize_crate_name(VARCHAR)
cratesfyi->                 RETURNS VARCHAR
cratesfyi->                 AS $$
cratesfyi$>                     SELECT LOWER(REPLACE($1, '_', '-'));
cratesfyi$>                 $$ LANGUAGE SQL;
CREATE FUNCTION
cratesfyi=> select normalize_crate_name('test-one');
ERROR:  function normalize_crate_name(unknown) does not exist
LINE 1: select normalize_crate_name('test-one');
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
cratesfyi=> select pg_temp.normalize_crate_name('test-one');
 normalize_crate_name 
----------------------
 test-one
(1 row)

cratesfyi=> select pg_temp.normalize_crate_name('test_one');
 normalize_crate_name 
----------------------
 test-one
(1 row)
```

Reply:
Yep, that's what I meant by using an explicit schema-name to call them. Which would mean having to change the sql run inside the app depending on whether it's in test mode or not.

Reply:
> Which would mean having to change the sql run inside the app depending on whether it's in test mode or not.

Ah, I see. Yeah @pietroalbini tried that a while back (I think in https://github.com/rust-lang/docs.rs/pull/556?) and it made it much more difficult to understand the code, so this looks fine.